### PR TITLE
`MaxPool` and `AveragePool` fix for `ceil_mode`

### PR DESCRIPTION
**Pull request by Hyunseok-Kim0**
### 2. Summary of corrections
Padding calculation is fixed for the case when `ceil_mode` is used.
Now extra padding value is added to match the output shape of tensorflow to ONNX. This extra padding is added to right, bottom side of feature map if needed.
However, `ceil_mode` caused bug in `AveragePool`. I guess it can be solved by using `average_multiplier` in this case, but couldn't fixed yet.

### 3. Before/After (If there is an operating log that can be used as a reference)

isnet convert result | isnet structure
---- | ----
 | 

Right, bottom side value difference in AveragePool when `ceil_mode` is True |
---- |
Screenshot from 2023-02-28 11-15-29

### 4. Issue number (only if there is a related issue)
#207

### CODE REVIEW

**Comment by Hyunseok-Kim0** [comment]
Now the bug is fixed and tflite poolings work same as pytorch pooling method.
I used code below to test for two inputs and random `ceil_mode` and `count_include_pad`.
If there is no problem after test, it is ready to be merged.

```python
import os
import shutil

import onnxruntime
import tensorflow as tf
import torch
import torchvision
from einops import rearrange
from torch import nn
import numpy as np
from onnx2tf import convert
import onnx_graphsurgeon as gs
import onnx


class dummy_network(nn.Module):

    def __init__(self, kernel=3, pad=1, stride=1, ceil_mode=False, count_include_pad=False):
        super(dummy_network, self).__init__()
        self.kernel = kernel
        self.pad = pad
        self.stride = stride
        self.ceil_mode = ceil_mode
        self.count_include_pad = count_include_pad

    def forward(self, x):

        if len(x.shape) == 3:
            result = torch.nn.AvgPool1d(self.kernel, self.stride, self.pad, ceil_mode=self.ceil_mode, count_include_pad=self.count_include_pad)(x)

        elif len(x.shape) == 4:
            result = torch.nn.AvgPool2d(self.kernel, self.stride, self.pad, ceil_mode=self.ceil_mode, count_include_pad=self.count_include_pad)(x)

        else:
            result = torch.nn.AvgPool3d(self.kernel, self.stride, self.pad, ceil_mode=self.ceil_mode, count_include_pad=self.count_include_pad)(x)

        return result


def main():
    os.makedirs("tflite", exist_ok=True)

    dummy_name = "dummy_avgpool"
    onnx_save_path = f"tflite/{dummy_name}.onnx"
    tflite_save_path = f"tflite/{dummy_name}_float32.tflite"

    if np.random.rand() > 0.5:
        dummy_x = torch.ones(1, 1, 32, 16, dtype=torch.float32)
    else:
        dummy_x = torch.ones(1, 1, 8, 11, dtype=torch.float32)
    np.random.seed(123)
    kernels = [np.random.randint(3, 8) for _ in range(500)]
    strides = [np.random.randint(2, k) for k in kernels]
    paddings = [np.random.randint(0, k / 2) for k in kernels]
    ceil_mode = True
    count_include_pad = True

    for k, s, p in zip(kernels, strides, paddings):
        # k, s, p = 9, 4, 4
        # k, s, p = 7, 6, 2
        # k, s, p = 7, 1, 2
        print("kernel, stride, padding : ", k, s, p)

        # randomly turn on count_include_pads
        if np.random.rand() > 1.5:
            count_include_pad = False

        if np.random.rand() > 1.5:
            ceil_mode = False

        model = dummy_network(kernel=k, stride=s, pad=p, ceil_mode=ceil_mode, count_include_pad=count_include_pad)
        model.eval()
        torch_output = model(dummy_x).numpy()

        model = dummy_network(kernel=k, stride=s, pad=p, ceil_mode=ceil_mode, count_include_pad=False)
        torch.onnx.export(model,
                          args=(dummy_x, ),
                          f=onnx_save_path,
                          input_names=["x"],
                          output_names=["result"],
                          opset_version=11)

        if count_include_pad:
            graph = gs.import_onnx(onnx.load(onnx_save_path))

            for i, gn in enumerate(graph.nodes):
                graph.nodes[i].attrs['count_include_pad'] = 1

            graph.toposort()
            onnx.save(gs.export_onnx(graph), onnx_save_path)

        # get onnx output
        # -----------------------------------------------------------------------------------------------
        onnx_session = onnxruntime.InferenceSession(onnx_save_path, providers=['CPUExecutionProvider'])
        onnx_inputs = dict(x=dummy_x.numpy())
        onnx_output = onnx_session.run(None, onnx_inputs)[0]

        # from onnx.reference import ReferenceEvaluator
        # ref_sess = ReferenceEvaluator(onnx_save_path, verbose=1)
        # results = ref_sess.run(None, onnx_inputs)

        convert(onnx_save_path, output_folder_path="tflite", check_onnx_tf_outputs_elementwise_close_full=True)
        # shutil.move(temp_tflite, tflite_save_path)

        # get tflite output
        # -----------------------------------------------------------------------------------------------
        tflite_onnx2tf = tf.lite.Interpreter(model_path=tflite_save_path)
        tflite_onnx2tf.allocate_tensors()

        if len(dummy_x.shape) == 3:
            tflite_onnx2tf.set_tensor(tflite_onnx2tf.get_input_details()[0]['index'],
                                      rearrange(dummy_x.numpy(), "n c h -> n h c"))
        elif len(dummy_x.shape) == 4:
            tflite_onnx2tf.set_tensor(tflite_onnx2tf.get_input_details()[0]['index'],
                                      rearrange(dummy_x.numpy(), "n c h w -> n h w c"))
        elif len(dummy_x.shape) == 5:
            tflite_onnx2tf.set_tensor(tflite_onnx2tf.get_input_details()[0]['index'],
                                      rearrange(dummy_x.numpy(), "n c d h w -> n d h w c"))

        tflite_onnx2tf.invoke()
        tflite_output = [tflite_onnx2tf.get_tensor(i['index']) for i in tflite_onnx2tf.get_output_details()][0]

        if len(dummy_x.shape) == 3:
            tflite_output = rearrange(tflite_output, "n h c -> n c h")
        elif len(dummy_x.shape) == 4:
            tflite_output = rearrange(tflite_output, "n h w c -> n c h w")
        elif len(dummy_x.shape) == 5:
            tflite_output = rearrange(tflite_output, "n d h w c -> n c d h w")

        if not ceil_mode:
            np.testing.assert_allclose(tflite_output, onnx_output, rtol=0.1, atol=1e-3)
        else:
            np.testing.assert_allclose(tflite_output, torch_output, rtol=0.1, atol=1e-3)

        print("convert done")

    return


if __name__ == "__main__":
    main()

```

**Comment by PINTO0309** [comment]
The 180x320 model with the largest shape discrepancy matched the output perfectly. Thanks! :smiley: 
image
